src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs

```diff
@@ -0,0 +1,443 @@
+                        nameof(actual));
+
+                if (Type != indexDef.Type)
+                    throw new ArgumentException(
+                        $"Expected index {Name} to be have Type='{Type}', got Type='{indexDef.Type}' instead",
+                        nameof(actual));
+
+                if (StartIndex != indexDef.StartIndex)
+                    throw new ArgumentException(
+                        $"Expected index {Name} to have StartIndex='{StartIndex}', got StartIndex='{indexDef.StartIndex}' instead",
+                        nameof(actual));
+
+                if (Count != indexDef.Count)
+                    throw new ArgumentException(
+                        $"Expected index {Name} to have Count='{Count}', got Count='{indexDef.Count}' instead",
+                        nameof(actual));
+            }
+
+            public override void Validate()
+            {
+                base.Validate();
+
+                if (StartIndex < 0)
+                    throw new ArgumentOutOfRangeException(nameof(StartIndex), "StartIndex cannot be negative");
+            }
+
+            public static IndexDef ReadFrom(ByteStringContext context, byte* location, int size)
+            {
+                var input = new TableValueReader(location, size);
+                var indexDef = new IndexDef();
+
+                byte* currentPtr = input.Read(1, out int currentSize);
+                indexDef.StartIndex = *(int*)currentPtr;
+
+                currentPtr = input.Read(2, out currentSize);
+                indexDef.Count = *(int*)currentPtr;
+
+                currentPtr = input.Read(3, out currentSize);
+                indexDef.IsGlobal = Convert.ToBoolean(*currentPtr);
+
+                currentPtr = input.Read(4, out currentSize);
+                Slice.From(context, currentPtr, currentSize, ByteStringType.Immutable, out indexDef.Name);
+
+                return indexDef;
+            }
+        }
+
+        public class DynamicKeyIndexDef : AbstractTreeIndexDef
+        {
+            public override TreeIndexType Type => TreeIndexType.DynamicKeyValues;
+
+            public delegate ByteStringContext.Scope IndexEntryKeyGenerator(ByteStringContext context, ref TableValueReader value, out Slice slice);
+
+            public IndexEntryKeyGenerator GenerateKey;
+
+            public override ByteStringContext.Scope GetValue(ByteStringContext context, ref TableValueReader value,
+                out Slice slice)
+            {
+                return GenerateKey(context, ref value, out slice);
+            }
+
+            public override ByteStringContext.Scope GetValue(ByteStringContext context, TableValueBuilder value,
+                out Slice slice)
+            {
+                using (context.Allocate(value.Size, out var buffer))
+                {
+                    value.CopyTo(buffer.Ptr);
```

Comment:
We need to see if we can optimize this. We are forcing a copy of the value here.
A better option would be to have two delegates, or pass both in.

Reply:
created a separate ticket for optimization 
https://issues.hibernatingrhinos.com/issue/RavenDB-18105